GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRow.cs

```diff
@@ -214,5 +217,57 @@ public void MoveLanesRight(int fromLane)
                 ++_segmentLanes[segment];
             }
         }
+
+        public int LaneCountAfterMovingLanesRight(int laneIndex)
+        {
+            if (laneIndex >= _laneCount)
+            {
+                return _laneCount;
+            }
+
+            if (_gaps?.Any(g => g > laneIndex) == true)
+            {
+                return _laneCount;
+            }
+
+            return _laneCount + 1;
+        }
+
+        /// <summary>
+        /// If this row is the parent row of <paramref name="segment"/>, returns the segment leading
+        /// to this row's first parent (if any). Otherwise, returns <paramref name="segment"/>.
+        /// </summary>
+        public RevisionGraphSegment FirstParentOrSelf(RevisionGraphSegment segment)
```

Comment:
I guess it should rather be `SelfOrFirstParent` in order to keep the segment straight as far as possible.
Only if it ends, take the first parent a continuation.

Reply:
The first parent should only be chosen for the first child. If coming from another child, the segment should not be continued.

Reply:
> I guess it should rather be `SelfOrFirstParent` in order to keep the segment straight as far as possible. Only if it ends, take the first parent as continuation.

Are you arguing for just a name change, or for a functionality change? `SelfOrFirstParent` and `FirstParentOrSelf` are essentially the same, since the segment cannot both continue *and* have a parent in this row. But I opted for the latter since it returns self also in the case where the segment ends and has no parent segment.

> The first parent should only be chosen for the first child.

I considered this. But how do you define the 'first' child segment? By order in the `Segments` list? By lane order in the row above (is this the same?)? By distance to the row of the child revision? In each case, I think I can find an example where this would lead to less straightening than desired.

Reply:
> `SelfOrFirstParent` and `FirstParentOrSelf` are essentially the same, since the segment cannot both continue *and* have a parent in this row. But I opted for the latter since it returns self also in the case where the segment ends and has no parent segment.

👍 Thank you for the explanation. I agree. 
(From just reading `if (segment.Parent != Revision)`, I do not get the meaning without looking into the docs of all these properties. Let's think about inversion and comments later when we have an agreed implementation.)  

I was just dropping guesses how to avoid the yellow segment being bent one lane to the right unnecessarily early (resulting in the free lane under the mouse cursor).
The cyan branch is branched from and back to the yellow branch - at least as I read the graph.
The cyan branch should not "attract" the yellow branch (although it would save one bend for the cyan branch without adding one to the yellow branch).

> I considered this. But how do you define the 'first' child segment? By order in the Segments list? **By lane order in the row above**?

👍
The yellow branch is the "first", the cyan one is the "second" and should not be continued with the ancestor.

> (is this the same?)

I guess yes. Though this is not guaranteed. It depends on the sorting algorithm. I am experimenting with an (immature) alternative in my private build.
Using `GetLaneIndexForSegment()` will be safe and shouldn't be too expensive.

![image](https://user-images.githubusercontent.com/36601201/223530026-7990732a-9002-47ea-82f7-41e30e892e8c.png)  ![image](https://user-images.githubusercontent.com/36601201/223850252-d919d94e-4fb1-49dc-b29e-4bc90aaf6621.png)

Reply:
Sorry, the above seems to be not caused by this PR. The yellow and the cyan segments share the same lane since they have the same parent.
Since `StraightenLanes` operates on single segments, it cannot be detected / avoided not easily.
So I am dropping this part of the request.